FEATURE: make packages sortable in package management BE module

crydotsnake opened the pull request:
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

**Upgrade instructions**

_None_

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

This pull request is related too: https://github.com/neos/flow-development-collection/pull/3363

There we decided to list the packages by their loading order in the `flow:package:list` command.

This pull request takes care that the packages are listed by the loading order in the backend module too.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

crydotsnake committed:
TASK: list packages by loading order in BE Module

Sebobo commented:
The module should then show an information about the order.

Btw. I usually prefer alphabetical order as customers tend to show me that list when we discuss updates.
Via screenshare the alphabetical order is more useful for me than the loading order.

Probably better to ask the community what they think or if people actually don't care.

Or make it switchable.

crydotsnake commented:
Personally i would be fine to sort the package by alphabetical order

Wdyt? @mhsdesign @kitsunet @ahaeslich

ahaeslich commented:
If we have use cases where either the alphabetical or loading order is needed, I would go with the option "make it sortable". As in "let the user decide which order is needed now". 

If we decide to have this selection in the backend module: how can we be sure the alphabetical order isn't needed by someone in the CLI (see: https://github.com/neos/flow-development-collection/pull/3363)? Is there any need to remove it there?

crydotsnake commented:
How do we continue here? Should we sort it by alphabetical order by default? or give the editor the option to also sort it by loading order?

crydotsnake commented:
@Sebobo Maybe i could solve it the same way as you did it in: https://github.com/neos/neos-development-collection/pull/5209

So that there would be a dropdown selector with the two options for example:

- Sort by Alphabetical order
- Sort by loading order.

Wdyt?

Sebobo commented:
Allow clicking on the headers to sort the table with alphabetical being the default plz :)

crydotsnake commented:
> Allow clicking on the headers to sort the table with alphabetical being the default plz :)

@Sebobo So now dropdown selector? and only by alphabetical order or with a another click by loading order too?

grebaldi committed:
BUGFIX: Convert *DateTime XML elements to \DateTimeImmutable during import

reissigMR committed:
[FIX] Find nodes marked as removed in other dimension

weblate committed:
TASK: Update translation files

Updated by 'Cleanup translation files' hook in Weblate.

Co-authored-by: Hosted Weblate <hosted@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/neos/neosmediabrowser-main-84/
Translation: Neos/Neos.Media.Browser - Main - 8.4

Sebobo committed:
BUGFIX: Allow boolean values for hidden properties in nodetype schema

This prevents matching warnings in the configuration module or when verifying nodetypes via cli.

Resolves: 5520

dos-format-d committed:
bugfix: missing flash messages after Dropzone upload in Neos Media Browser

dlubitz committed:
Merge pull request #5523 from dos-format-d/bugfix/flash-messages-not-shown-after-upload

BUGFIX: Missing flash messages after Dropzone upload in Neos Media Browser

mhsdesign committed:
Merge pull request #5521 from neos/bugfix/hidden-property

BUGFIX: Allow boolean values for hidden properties in nodetype schema

mhsdesign committed:
BUGFIX: Boot media module flash-messages correctly if rendered outside Neos module

Previously only flashmessages in the Default template (e.g. on show) where shown but neither for editing or uploading

nezaniel committed:
Merge pull request #5528 from mhsdesign/bugfix/ensure-flashmessages-work-for-media-browser-if-rendered-outside-neos-module

BUGFIX: Boot media module flash-messages correctly if rendered outside Neos module

kitsunet committed:
Merge pull request #5498 from beromir/fix/find-removed-nodes-in-other-dimension

[BUGFIX] Find nodes marked as removed in other dimension

Commit:
TASK: Add changelog for 8.3.23 [skip ci]

See https://jenkins.neos.io/job/neos-release/464/

dlubitz committed:
TASK: Allow neos/fusion-form in version 3.x to support neos/symfonymailer.

so-grimm committed:
Bugfix/5373: Fix publishing for `Neos.Neos:RestrictedEditor` from 3rd level workspace

- `WorkspacesController`: explicitly define the target workspace for case `publish` in `publishOrDiscardNodesAction()`

Issue: https://github.com/neos/neos-development-collection/issues/5373

so-grimm committed:
Bugfix/5373: adjust indentation for comments to fix styleCI issue

Issue: https://github.com/neos/neos-development-collection/issues/5373

dlubitz committed:
Merge pull request #5546 from queoGmbH/bugfix/5373_restrictedEditor-publishing-childNodes-from-3rd-level-workspace

BUGFIX: Fix publishing for `Neos.Neos:RestrictedEditor` from 3rd level workspace

dlubitz committed:
TASK: Remove support for neos/fusion-form 1.x

dlubitz committed:
Merge pull request #5544 from dlubitz/task/fusion-form-version

TASK: Allow neos/fusion-form in version 3.x to support neos/symfonymailer.

dlubitz committed:
BUGFIX: Replace skip migrations with simple return

kitsunet committed:
Merge pull request #5555 from dlubitz/bugfix/replace-skipif-migrations

BUGFIX: Replace skip migrations with simple return

dlubitz committed:
Merge branch '8.3' into 8.4

Commit:
TASK: Add changelog for 8.3.24 [skip ci]

See https://jenkins.neos.io/job/neos-release/468/

mhsdesign committed:
BUGFIX: Boot media module `NeosCMS.I18n.translate()` correctly if rendered outside Neos module

This bug was uncovered because translate() simply returns undefined, if not initialised. No error no fallback, nuthing.

Sebobo committed:
BUGFIX: Convert selectbox group labels i18n values

markusguenther committed:
Merge pull request #5560 from neos/bugfix/translate-selectbox-group-labels

BUGFIX: Convert selectbox group labels i18n values

Benjamin-K committed:
FIX: Correctly check workspaces in NodeData::move()

mhsdesign committed:
Merge pull request #5553 from neos/bugfix/mediabrowser-upload-action-file-too-big-message-is-empty

BUGFIX: MediaBrowser's upload action shows empty flash message if file is too big

dlubitz committed:
Merge pull request #5562 from die-wegmeister/fix/8.3-move-nodes-workspace-check

Fix checking workspaces when move nodes

dlubitz committed:
Merge remote-tracking branch 'origin/8.3' into 8.4

skurfuerst committed:
Merge pull request #5249 from grebaldi/bugfix/5248/date-time-during-site-import

BUGFIX: Convert *DateTime XML elements to \DateTimeImmutable during import

Sebobo commented:
How about adding the loading order sequence number in its own column, then you can sort via the header like we do in the user module?

crydotsnake commented:
> How about adding the loading order sequence number in its own column, then you can sort via the header like we do in the user module?

I already started working on that like i did it in the user management module. If you want we can take a look at it tomorrow so i can show you the current state :)

Commit:
FEATURE: make package list sortable in package management backend module

crydotsnake commented:
Since its now possible to sort the list, how should we change the sorting in the CLI command?:

https://github.com/neos/flow-development-collection/pull/3363

the same way as here? so rather ASC, or DESC?

bwaidelich commented on Neos.Neos/Classes/Controller/Module/Administration/PackagesController.php:
```diff
@@ -30,13 +31,15 @@ class PackagesController extends AbstractModuleController
      * @return void
      */
-    public function indexAction()
+    public function indexAction(string $sortBy = 'name', string $sortDirection = QueryInterface::ORDER_ASCENDING)
```
I suggest to avoid introducing this dependency to the Persistence submodule

```suggestion
    public function indexAction(string $sortBy = 'name', bool $descending = false)
```

bwaidelich commented on Neos.Neos/Classes/Controller/Module/Administration/PackagesController.php:
```diff
@@ -46,13 +49,23 @@ public function indexAction()
+                $valueB = $b[$sortBy] ?? '';
+
+                $result = strnatcasecmp((string)$valueA, (string)$valueB);
+                return $sortDirection === QueryInterface::ORDER_DESCENDING ? -$result : $result;
```
```suggestion
                return $sortDirection === $descending ? -$result : $result;
```

bwaidelich commented on Neos.Neos/Classes/Controller/Module/Administration/PackagesController.php:
```diff
@@ -46,13 +49,23 @@ public function indexAction()
             'packageGroups' => $packageGroups,
-            'isDevelopmentContext' => $this->objectManager->getContext()->isDevelopment()
+            'isDevelopmentContext' => $this->objectManager->getContext()->isDevelopment(),
+            'sortDirection' => $sortDirection,
```
```suggestion
            'descending' => $descending,
```

bwaidelich commented on Neos.Neos/Resources/Private/Templates/Module/Administration/Packages/Index.html:
```diff
@@ -6,13 +6,45 @@
+			<tr>
+				<th title="{neos:backend.translate(id: 'packages.sortBy.packageName', source: 'Modules', package: 'Neos.Neos')}">
+					<f:render section="headerLabel" arguments="{
+						sortDirection: sortDirection,
```
```suggestion
						descending: descending,
```

bwaidelich commented on Neos.Neos/Resources/Private/Templates/Module/Administration/Packages/Index.html:
```diff
@@ -55,3 +87,29 @@
+<f:section name="headerLabel">
+	<f:if condition="{sortBy} === {sortingField}">
+		<f:then>
+			<f:if condition="{sortDirection} === 'ASC'">
+				<f:then>
+					<f:link.action action="index" arguments="{sortDirection: 'DESC', sortBy: sortingField}" addQueryString="TRUE">
+						<neos:backend.translate id="{labelIdentifier}" source="Modules" package="Neos.Neos"/>
+						<i class="fas fa-sort-amount-up"></i>
+					</f:link.action>
+				</f:then>
+				<f:else>
+					<f:link.action action="index" arguments="{sortDirection: 'ASC', sortBy: sortingField}" addQueryString="TRUE">
+						<neos:backend.translate id="{labelIdentifier}" source="Modules" package="Neos.Neos"/>
+						<i class="fas fa-sort-amount-down"></i>
+					</f:link.action>
+				</f:else>
+			</f:if>
```
```suggestion
			<f:if condition="{descending}">
				<f:then>
					<f:link.action action="index" arguments="{sortDirection: 'ASC', sortBy: sortingField}" addQueryString="TRUE">
						<neos:backend.translate id="{labelIdentifier}" source="Modules" package="Neos.Neos"/>
						<i class="fas fa-sort-amount-down"></i>
					</f:link.action>
				</f:then>
				<f:else>
					<f:link.action action="index" arguments="{sortDirection: 'DESC', sortBy: sortingField}" addQueryString="TRUE">
						<neos:backend.translate id="{labelIdentifier}" source="Modules" package="Neos.Neos"/>
						<i class="fas fa-sort-amount-up"></i>
					</f:link.action>
				</f:else>
			</f:if>
```

bwaidelich commented on Neos.Neos/Resources/Private/Translations/en/Modules.xlf:
```diff
@@ -489,6 +489,21 @@
 				<source>Package Type</source>
 			</trans-unit>
+      <trans-unit id="packages.version" xml:space="preserve">
+				<source>Package Version</source>
```
This is too long and breaks the header on smaller screens. We could make the version column larger, but IMO it's easier to shorten the header

```suggestion
				<source>Version</source>
```

bwaidelich commented:
I tested this and in basically works as intended even though I really don't know why I should want to sort packages by their version.

Also, regarding loading order: The loading order can be an important indicator for bugs (e.g. broken policies if someone defined privileges/roles without declaring dependency to `neos/flow`).

Adding a sortable loading order column works in theory:
![image](https://github.com/user-attachments/assets/ffd691c3-fa55-4bf7-9c5d-f1aff8559ba5)

But IMO this is far too low level for anyone to understand. And the groups render it practically useless.. 
A better solution would IMO be a detail view for the packages that show the dependency graph. But I guess, that's out of scope for this.

Also, didn't we decide to merge the package module into `Configuration`?

Sebobo commented:
We decided to not include this in 8.4 and target 9.1 instead and convert the changes into the HTMX based module we started working on during contribution day if we can make it in time.

Sebobo reviewed (changes requested):
Needs rebase on 9.1
